modules/personalize_taxonomy_context/personalize_taxonomy_context.module

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * @file personalize_taxonomy_context.module
+ * Provides a visitor context plugin for taxonomy.
+ */
+
+/**
+ * Implements hook_personalize_visitor_contexts().
+ */
+function personalize_taxonomy_context_personalize_visitor_context() {
+  $info = array();
+  $path = drupal_get_path('module', 'personalize_taxonomy_context') . '/plugins';
+  $info['taxonomy_context'] = array(
+    'path' => $path . '/visitor_context',
+    'handler' => array(
+      'file' => 'TaxonomyContext.inc',
+      'class' => 'TaxonomyContext',
+    ),
+  );
+  return $info;
+}
+
+/**
+ * Implements hook_node_view().
+ */
+function personalize_taxonomy_context_node_view($node, $view_mode, $langcode) {
+  personalize_taxonomy_context_node_taxonomy($node, $view_mode == 'full');
+}
+
+/**
+ * Handles the node-specific taxonomy data.
+ *
+ * @param $node
+ *   (optional) If passed then the taxonomy terms for the selected node
+ *   are added to the page taxonomy context.
+ * @param bool $primary
+ *   True if this is the primary node displayed for the page, false otherwise.
+ * @return array
+ *   Returns the taxonomy context for the page.
+ */
+function personalize_taxonomy_context_node_taxonomy($node = NULL, $primary = FALSE) {
+  $taxonomy = &drupal_static(__FUNCTION__);
+  if (!isset($taxonomy)) {
+    $taxonomy = array();
+  }
+
+  // Add vocabulary and terms only if:
+  // 1) Has a node, and
+  // 2) The node is the primary node displayed for the page.
+  if (!isset($node) || !$primary) {
+    return $taxonomy;
+  }
+
+  // For each vocabulary, add any terms that are applied to the viewed node.
+  $vocabularies = array();
+  foreach($node as $property_name => $property_value) {
+    // If the field is not a vocabulary, move on.
+    if (!isset($property_value['und'])) {
+      continue;
+    }
+    // Process vocabulary.
+    foreach ($property_value['und'] as $term) {
```

Comment:
At this point you won't actually be sure you have taxonomy terms so I would rename these variables such that this loop reads `foreach ($property_value[$field_lang] as $field)` and then below you are checking whether they are in fact taxonomy term fields.